redolog/redolog_manager_master.go

```diff
@@ -0,0 +1,149 @@
+	managers map[string]map[int]RedologManager
+}
+
+// NewRedoLogManagerMaster create RedoLogManagerMaster instance
+func NewRedoLogManagerMaster(c *common.RedoLogConfig, diskStore diskstore.DiskStore, metaStore metastore.MetaStore) (*RedoLogManagerMaster, error) {
+	return NewKafkaRedoLogManagerMaster(c, diskStore, metaStore, nil)
+}
+
+// NewKafkaRedoLogManagerMaster convenient function if the kafka consumer can be passed in from outside
+func NewKafkaRedoLogManagerMaster(cfg *common.RedoLogConfig, diskStore diskstore.DiskStore, metaStore metastore.MetaStore, consumer sarama.Consumer) (*RedoLogManagerMaster, error) {
+
+	if cfg == nil {
+		cfg = &common.RedoLogConfig{}
+	}
+	if cfg.KafkaConfig.Enabled {
+		if len(cfg.KafkaConfig.Brokers) == 0 {
+			return nil, fmt.Errorf("No kafka broker info configured")
+		}
+		if consumer == nil {
+			var err error
+			if consumer, err = sarama.NewConsumer(cfg.KafkaConfig.Brokers, sarama.NewConfig()); err != nil {
+				return nil, err
+			}
+		}
+	} else {
+		consumer = nil
+	}
+
+	return &RedoLogManagerMaster{
+		RedoLogConfig:      cfg,
+		diskStore:          diskStore,
+		managers:           make(map[string]map[int]RedologManager),
+		metaStore:          metaStore,
+		consumer:           consumer,
+	}, nil
+}
+
+// NewRedologManager create compositeRedoLogManager on specified table/shard
+// each table/shard should only have one compositeRedoLogManager
+func (m *RedoLogManagerMaster) NewRedologManager(table string, shard int, tableConfig *metaCom.TableConfig) (RedologManager, error) {
+	utils.GetLogger().With("action", "ingestion", "table", table, "shard", shard).Info("Create Redolog Manager")
+	m.Lock()
+	defer m.Unlock()
+
+	var tableManager map[int]RedologManager
+	var ok bool
+	if tableManager, ok = m.managers[table]; !ok || tableManager == nil {
+		tableManager = make(map[int]RedologManager)
+		m.managers[table] = tableManager
+	}
+	manager, ok := tableManager[shard]
+	if ok && manager != nil {
+		return nil, fmt.Errorf("NewRedologManager for table: %s, shard: %d is already running", table, shard)
+	}
+
+	if m.RedoLogConfig.KafkaConfig.Enabled {
+		commitFunc := m.metaStore.UpdateRedoLogCommitOffset
+		checkPointFunc := m.metaStore.UpdateRedoLogCheckpointOffset
+		getCommitOffsetFunc := m.metaStore.GetRedoLogCommitOffset
+		getCheckpointOffsetFunc := m.metaStore.GetRedoLogCheckpointOffset
+
+		if m.RedoLogConfig.DiskConfig.Disabled {
+			manager = newKafkaRedoLogManager(m.RedoLogConfig.Namespace, table, shard, m.consumer, true, commitFunc, checkPointFunc, getCommitOffsetFunc, getCheckpointOffsetFunc)
+		} else {
+			manager = newCompositeRedoLogManager(m.RedoLogConfig.Namespace, table, shard, tableConfig, m.consumer, m.diskStore, commitFunc, checkPointFunc, getCommitOffsetFunc, getCheckpointOffsetFunc)
+		}
+	} else {
+		manager = newFileRedoLogManager(int64(tableConfig.RedoLogRotationInterval), int64(tableConfig.MaxRedoLogFileSize), m.diskStore, table, shard)
+	}
+
+	tableManager[shard] = manager
```

Comment:
I think you still need to assign tableManager back to m.managers
`m.managers[table] = tableManager`

Reply:
dont think so